### PR TITLE
Consolidated Install/Launch Scripts and Improve README

### PR DESCRIPTION
**Pull request by jstayco**
Install scripts have been consolidated for every non-Windows OS. Python requirements were consolidated. README improved to work locally (avoid Github links) and provide more clarity on how to install and launch everything with the newer scripts.

Notably, the **Linux support needs to be tested** to ensure I got the right commands and string detection for distributions. I have tested the macOS portion successfully for both installation and launching the GUI.

### CODE REVIEW

**Comment by jstayco** [comment]
#328 This is a good first step to improving the cross platform support story to include macOS.

**Comment by bmaltais** [comment]
Thank you for this nice PR. Can someone with a Linux system validate that all is working as intended?

File: setup.sh [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      echo "Python TK found! Skipping install!"
+    fi
+  elif "$distro" | grep -Eqi "Fedora|CentOS|Redhat"; then
+    if ! rpm -qa | grep -qi python3-tkinter; then
+      if [ root = true ]; then
+        dnf install python3-tkinter
+      else
+        echo "This script needs to be run as root or via sudo to install packages."
+        exit 1
+      fi
```
Comment (Galunid) [inline setup.sh:33]: Can we add ArchLinux/Manjaro support? I believe the command to install tkinter is `pacman -S tk`

**Review by Galunid — COMMENTED** [review]
Could we also include script to run on runpod? No idea if there's better way to test it, but what I've came up with:
```bash
if [[ $(ps -aux | grep "/sbin/docker-init -- /start.sh"  | grep -v grep | wc -l) -ge 1 ]];
then
  echo "probably running on runpod";
fi
```

Script I use, based on discussion in #379 
```bash
#!/bin/bash
VENV_DIR="/workspace/kohya_ss/venv"
BRANCH="dev"
if [ -d "$VENV_DIR" ]; then
  echo "Virtual Environment already exists";
  echo "Reinstalling system dependencies";
  apt update -y; apt install -y python3-tk
  export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/workspace/kohya_ss/venv/lib/python3.10/site-packages/tensorrt/
  export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/workspace/kohya_ss/venv/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/
  cd /workspace/kohya_ss
else
  echo "Installing from scratch";
  cd /workspace
  git clone https://github.com/bmaltais/kohya_ss.git
  cd kohya_ss
  git checkout "$BRANCH"
  sed -i 's/sudo apt install python3-tk/apt update -y; apt install -y python3-tk/g' ubuntu_setup.sh && sed -i "s/interface.launch(\*\*launch_kwargs)/interface.launch(\*\*launch_kwargs,share=True)/g" kohya_gui.py
  ./ubuntu_setup.sh
  source "$VENV_DIR/bin/activate"
  pip install tensorrt
  cd /workspace/kohya_ss/venv/lib/python3.10/site-packages/tensorrt
  ln -s libnvinfer_plugin.so.8 libnvinfer_plugin.so.7
  ln -s libnvinfer.so.8 libnvinfer.so.7
  cd /workspace/kohya_ss/venv/lib/python3.10/site-packages/nvidia/cuda_runtime/lib
  ln -s libcudart.so.12 libcudart.so.11.0
  export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/workspace/kohya_ss/venv/lib/python3.10/site-packages/tensorrt/
  export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/workspace/kohya_ss/venv/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/
fi
./gui.sh
```

I believe it's pretty useful, since a lot of people don't have strong enough GPU

File: setup.sh [hunk 2]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env bash
+
+# This file will be the host environment setup file for all operating systems other than base Windows.
+
+if [[ "$OSTYPE" == "linux-gnu"* ]]; then
+  # Check if root or sudo
+  root=true
+  if [ "$EUID" -ne 0 ]; then
+    root=false
```
[the remark this answers is not in the document]
Comment (jstayco) [reply setup.sh:9]: We exit at the install commands because the pip commands do not need root privileges. Some professional places require IT or admin action to install packages, but not pip. No need to exclude those users.

**Comment by jstayco** [comment]
I've never used runpod before, so forgive my ignorance, but it looks like your one-liner runpod detection just detects generic docker container. Is that intended and we want the same behavior in any docker container whether runpod is associated or not? I'm sure we could also do the steps you outlined on runpod detection, but some of those look potentially highly specific to your setup. Could you outline for me what needs to be done in order to properly run on runpod? For example, is tensorrt something runpod specific? Do we need to link in specific libs on specific versions of Ubuntu? If you can paint a clear picture for me of the steps we need to take maybe in a list form in plain English I'm happy to translate that into the script, so it's truly universal.

**Comment by jstayco** [comment]
I'm looking at the runpod stuff and it seems like it's just a kubernetes-type service allowing access to GPUs. If that's the case, I think you're better off modifying a DOCKERFILE to do the steps you wanted then to try to use a Bash script to do so. I can also start that up for you, but that would have to be a bit later as I'm focusing on macOS efforts for this go around. I advise keeping the runpod out of the bash setup scripts for now though if I'm correct about the docker container story.

**Comment by jstayco** [comment]
Ok. I think I get the steps @Galunid was doing earlier. I see the lib linking. So I'm thinking I can shim that into this setup.sh script, but I stand by my assessment for the need for a real DOCKERFILE. So I'll try to do both. I do not have an Nvidia GPU, so I have zero way to test a Dockerfile. So what I'll do is make a separate PR for that Dockerfile so it can be tested in isolatation. I can provide instructions on how to test build and run a Dockerfile as well if needed.

File: setup.sh (hunk 1 above)
Comment (jstayco) [reply setup.sh:81]: Latest commit should fix this for Manjaro specifically, but also any other "family" we add. For instance, I added the Ubuntu family as well which should cover Mint Linux, Pop OS, etc.

**Comment by jstayco** [comment]
@Galunid @bmaltais I admit this got a bit out of hand, but now this install script should install just about anywhere in any condition. Def needs runpod testing though because all I tried to do was fit the above script into our new script and make it configurable.

Runpod workflow should look like runpod pre-setup -> normal Linux OS setup that applies to everyone (so you'd hit the Ubuntu or Redhat clause there) -> run pod post-setup -> auto call gui.sh. And it should work even if you have your own private fork of kohya_ss or anything like that.

**Comment by bmaltais** [comment]
Would you say this is ready to merge as it stand right now?

**Comment by jstayco** [comment]
> 

I think it looks ready to me. Seemed to be fine on macOS and is certainly better than what is currently in place. But I'd feel more comfortable if @Galunid is able to confirm my runpod pieces are correct. If you want to merge it anyway, I can just keep an eye on the Issues page for any installation issues and handle those as they pop up.

File: setup.sh [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    if [ -d "$VENV_DIR" ]; then
+      echo "Pre-existing installation on a runpod detected."
+      export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:"$VENV_DIR"/lib/python3.10/site-packages/tensorrt/
+      export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:"$VENV_DIR"/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/
+      cd "$DIR" || exit 1
+      sed -i "s/interface.launch(\*\*launch_kwargs)/interface.launch(\*\*launch_kwargs,share=True)/g" ./kohya_gui.py
+    else
+      echo "Clean installation on a runpod detected."
+      cd "$BASE_DIR" || exit 1
+      git clone "$GIT_REPO"
```
Comment (Galunid) [inline setup.sh:130]: I haven't tested it yet, but this will most likely fail. 
```bash
$ DIR="/workspace/kohya_ss";
$ BASE_DIR="$(echo "$DIR" | cut -d "/" -f2)";
$ VENV_DIR="$DIR/venv";
$ echo "D: $DIR, BD: $BASE_DIR, VD: $VENV_DIR"
D: /workspace/kohya_ss, BD: workspace, VD: /workspace/kohya_ss/venv
```
on runpod we start in `/workspace/stable-diffusion/` usually (depends on template). So `$BASE_DIR` needs to be absolute path.
We can skip cloning repo, altogether since if we are running this script we most likely already are inside one. Maybe we can check if `/workspace/kohya_ss` exists and if not then clone there since we depend on this path. Or a simple solution, if we are in git repo, then `DIR="$(pwd)"`, `VENV_DIR="$DIR/venv"`

**Comment by Galunid** [comment]
Sorry @jstayco I missed last comment yesterday and didn't know you plan on working on contenerization. I started playing a bit with Dockerfile for runpod since I didn't know you can use it with custom images. I got something very basic that seemed to run, although I didn't check if training actually works. I can test it if you want. 
```Dockerfile
FROM ubuntu:22.04

SHELL ["/bin/bash", "-c"] # because "source" command is broken in dash/sh
WORKDIR /workspace

COPY . .

RUN apt-get update -y
RUN apt-get install -y python3-tk python3.10-venv

RUN python3 -m venv venv
RUN source venv/bin/activate && pip install torch==1.12.1+cu116 torchvision==0.13.1+cu116 --extra-index-url https://download.pytorch.org/whl/cu116 && pip install --use-pep517 --upgrade -r requirements.txt && pip install -U -I --no-deps https://github.com/C43H66N12O12S2/stable-diffusion-webui/releases/download/linux/xformers-0.0.14.dev0-cp310-cp310-linux_x86_64.whl
RUN source venv/bin/activate && pip install tensorrt

RUN ln -s venv/lib/python3.10/site-packages/tensorrt/libnvinfer_plugin.so.8 venv/lib/python3.10/site-packages/tensorrt/libnvinfer_plugin.so.7
RUN ln -s venv/lib/python3.10/site-packages/tensorrt/libnvinfer.so.8 venv/lib/python3.10/site-packages/tensorrt/libnvinfer.so.7
RUN ln -s venv/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/libcudart.so.12 venv/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/libcudart.so.11.0

RUN export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/workspace/venv/lib/python3.10/site-packages/tensorrt/
RUN export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/workspace/venv/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/

CMD ["./gui.sh", "--listen", "0.0.0.0"]
EXPOSE 7860
```

File: setup.sh [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  -d DIR, --dir=DIR             The full path you want kohya_ss installed to.
+  -g, --git_repo                You can optionally provide a git repo to checkout for runpod installation. Useful for custom forks.
+  -r, --runpod                  Forces a runpod installation. Useful if detection fails for any reason.
+  -h, --help                    Show this screen.
+EOF
+}
+
+# Variables defined before the getopts loop, so we have sane default values.
+DIR="/workspace/kohya_ss"
+BRANCH="dev"
```
[the remark this answers is not in the document]
Comment (jstayco) [reply setup.sh:27]: Good call. I just went with your script. Will fix.

File: setup.sh (hunk 3 above)
Comment (jstayco) [reply setup.sh:130]: Agree with your DIR point. I'll add some more logic to better detect and make fewer assumptions. As it stands it really only works if you're one folder deep.

**Comment by jstayco** [comment]
@Galunid Check out the latest commit. I think that answers your concerns and makes it more resilient. Also, I really appreciate you taking the time to deep dive this before the merge.

File: setup.sh [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+      if [ ! -d "$DIR/.git" ]; then
+        echo "Cloning $GIT_REPO."
+        git clone "$GIT_REPO"
+        cd "$DIR" || exit 1
+        git checkout "$BRANCH"
+      else
+        cd "$DIR" || exit 1
+        echo "git repo detected. Attempting tp update repo instead."
+        echo "Updating: $GIT_REPO"
+        git pull "$GIT_REPO"
```
Comment (Galunid) [inline setup.sh:139]: I'd be hesitant to update repo every time, since we don't update dependencies etc. We also have `upgrade.sh` script.

**Comment by Galunid** [comment]
@jstayco tested on runpod
```bash
./setup.sh: line 60: env_var_exists: command not found
./setup.sh: line 60: env_var_exists: command not found
```

**Comment by jstayco** [comment]
Haha, yeah. Silly mistake; fixed!

File: setup.sh (hunk 5 above)
Comment (jstayco) [reply setup.sh:139]: So, I did that only if we detect a git repo in the configured installation folder, but we do not detect a venv folder. I'm assuming a busted or new environment at that point which is why I did it this way. The installation script down below should be installing the pip requirements.

**Comment by Galunid** [comment]
It also seems that python3-tk was not installed correctly on runpod, it works, after I installed it by hand. It logged 
```
Your branch is up to date with 'origin/master'.
Ubuntu detected.
Installing Python TK if not found on the system.
Python TK found! Skipping install!
```

File: setup.sh (hunk 5 above)
Comment (Galunid) [reply setup.sh:139]: My bad, I missed the check if venv doesn't exist.

**Comment by jstayco** [comment]
Yessir! Try pulling and running now.

**Comment by Galunid** [comment]
@jstayco 
```bash
remote: Total 2009 (delta 377), reused 436 (delta 319), pack-reused 1483
Receiving objects: 100% (2009/2009), 3.30 MiB | 20.60 MiB/s, done.
Resolving deltas: 100% (1341/1341), done.
Already on 'master'
Your branch is up to date with 'origin/master'.
Installing Python TK if not found on the system.
Ubuntu detected.
This script needs to be run as root or via sudo to install packages.
root@ce28438e7d33:/workspace/stable-diffusion-webui/kohya_ss#
```

**Comment by jstayco** [comment]
Interesting the check is `"$EUID" -ne 0`. Maybe run pod is doing something different there? I know plenty of containers on Openshift and such will randomize UIDs. I wonder if that's what that one is.

**Comment by Galunid** [comment]
@jstayco root detection worked this time, it seems python3-tk was installed, now it's installing the rest of the dependencies, I'll let you know if it works soon

**Comment by Galunid** [comment]
```log
ln: failed to create symbolic link '/workspace/kohya_ss/venv/lib/python3.10/site-packages/tensorrt/libnvinfer_plugin.so.7': No such file or directory
ln: failed to create symbolic link '/workspace/kohya_ss/venv/lib/python3.10/site-packages/tensorrt/libnvinfer.so.7': No such file or directory
ln: failed to create symbolic link '/workspace/kohya_ss/venv/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/libcudart.so.11.0': No such file or directory
```

**Comment by jstayco** [comment]
> ln: failed to create symbolic link '/workspace/kohya_ss/venv/lib/python3.10/site-packages/tensorrt/libnvinfer_plugin.so.7': No such file or directory
> ln: failed to create symbolic link '/workspace/kohya_ss/venv/lib/python3.10/site-packages/tensorrt/libnvinfer.so.7': No such file or directory
> ln: failed to create symbolic link '/workspace/kohya_ss/venv/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/libcudart.so.11.0': No such file or directory

Take a look at setup.sh lines 214-219. Believe I took that from your script. Did I mix up the source and targets?

**Comment by Galunid** [comment]
```
root@d495df2db040:/workspace/kohya_ss/venv/lib/python3.10/site-packages# cd tensorrt
bash: cd: tensorrt: No such file or directory
```
seems `pip install tenssort` is missing somewhere

**Comment by jstayco** [comment]
Very curious since it is all in one code block. That means it's hitting the logic, but tenssort didn't get installed?
```    pip install tensorrt
    ln -s "$VENV_DIR/lib/python3.10/site-packages/tensorrt/libnvinfer_plugin.so.8" \
      "$VENV_DIR/lib/python3.10/site-packages/tensorrt/libnvinfer_plugin.so.7"
    ln -s "$VENV_DIR/lib/python3.10/site-packages/tensorrt/libnvinfer.so.8" \
      "$VENV_DIR/lib/python3.10/site-packages/tensorrt/libnvinfer.so.7"
    ln -s "$VENV_DIR/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/libcudart.so.12" \
      "$VENV_DIR/lib/python3.10/site-packages/nvidia/cuda_runtime/lib/libcudart.so.11.0"
```

**Comment by jstayco** [comment]
> My bad, I didn't manually increase container volume size and it run out of space and failed to install tensorrt

Should I add a free space check for runpod/docker environments?

**Comment by Galunid** [comment]
Perhaps we could add a warning to set container disk to at least 10GB, I caught myself forgetting about it twice, usually the sign sth went wrong was tensorrt not being installed. On a happier note, *almost* everything went fine. It didn't expose public url (sed -i ... share=True). Also `./gui.sh` was run before `accelerate config`, which will most likely cause problems

**Comment by jstayco** [comment]
Now we're fancy! Not only do you have your 10Gb warning, but we now have interactive and non-interactive installation methods. The only real difference (for now?) is that we have an accelerate config file in the repo that we copy to the expected location based on the HuggingFace guidance. Non interactive (default)? We just check to see if those files exist and copy our default config right in if no file is found. We don't overwrite existing configs. If you run it with -i or --interactive then it will prompt for access config like it always did. 

@bmaltais I have our accelerate configuration file in a $project/config_files/accelerate folder. I thought that would make it obvious without cluttering the repo. Please let me know if you'd like that moved elsewhere. The values for this configuration file were taken from my machine where I just followed your README, so let me know if any of those values need changed for our default installs.

**Comment by bmaltais** [comment]
@jstayco that location sound good to me.

**Comment by Galunid** [comment]
```
df: /workspace/kohya_ss: No such file or directory
```
On a setup with Container disk 10GB and 35GB volume disk which should be enough to have kohya
```
root@82317a39e655:/workspace/stable-diffusion-webui/kohya_ss# df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay          10G   63M   10G   1% /
tmpfs            64M     0   64M   0% /dev
tmpfs            63G     0   63G   0% /sys/fs/cgroup
shm              20G     0   20G   0% /dev/shm
/dev/nvme0n1p2   49G   15G   33G  32% /usr/bin/nvidia-smi
/dev/nvme0n1p3   35G  9.7G   26G  28% /workspace
tmpfs            63G   12K   63G   1% /proc/driver/nvidia
tmpfs            63G  4.0K   63G   1% /etc/nvidia/nvidia-application-profiles-rc.d
tmpfs            13G  9.2M   13G   1% /run/nvidia-persistenced/socket
tmpfs            63G     0   63G   0% /proc/asound
tmpfs            63G     0   63G   0% /proc/acpi
tmpfs            63G     0   63G   0% /proc/scsi
tmpfs            63G     0   63G   0% /sys/firmware
```

Also
```
./setup.sh: line 69: declare: HF_HOME: not found
./setup.sh: line 69: declare: XDG_CACHE_HOME: not found
cp: cannot stat './config_files/accelerate/default_config.yaml': No such file or directory
Validating that requirements are satisfied.
All requirements satisfied.
Load CSS...
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
```
To access pod outside docker we need public link (share=True in launch()), so that part doesn't work

**Comment by jstayco** [comment]
@Galunid I think I hit all your points. Instead of directly modifying the kohya_gui.py file to expose public URL I'm attempting to just use the vehicle provided and call "gui.sh --share" per the README's guidance. That should pass right through if you use the -p or --public switch on this script now.

**Comment by Galunid** [comment]
I'm getting
```
cp: cannot stat '/workspace/kohya_ss/config_files/accelerate/default_config.yaml': No such file or directory
```
also
```
root@82ae62b9b7b9:/workspace/stable-diffusion-webui/kohya_ss# ./setup.sh -p
./setup.sh: illegal option -- p
```
`--public`  works though (and yes, the instance can be accessed!)

**Comment by jstayco** [comment]
> I'm getting
> 
> ```
> cp: cannot stat '/workspace/kohya_ss/config_files/accelerate/default_config.yaml': No such file or directory
> ```
> 
> also
> 
> ```
> root@82ae62b9b7b9:/workspace/stable-diffusion-webui/kohya_ss# ./setup.sh -p
> ./setup.sh: illegal option -- p
> ```
> 
> `--public` works though (and yes, the instance can be accessed!)

If you run the script with `-d /workspace/stable-diffusion-webui/kohya_ss` does that config error go away? I think it may be because you are using the default install location which is `/workspace/kohya_ss` instead of your real one. I'm pretty sure I got the -p issue as well. 

And no problem on the way; I just appreciate you heading off hundreds of user issues 😄

File: setup.sh [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+      echo "Clean installation on a runpod detected."
+      cd "$PARENT_DIR" || exit 1
+      if [ ! -d "$DIR/.git" ]; then
+        echo "Cloning $GIT_REPO."
+        git clone "$GIT_REPO"
+        cd "$DIR" || exit 1
+        git checkout "$BRANCH"
+      else
+        cd "$DIR" || exit 1
+        echo "git repo detected. Attempting tp update repo instead."
```
Comment (Galunid) [inline setup.sh:200]: typo `tp` -> `to`

**Comment by Galunid** [comment]
@jstayco Yes, the error is gone now that I set correct repo path. There's issue where when we rerun script it tries to reinstall all the dependencies and prints all are satisfied and then launches. I think it should first check if venv exist and if so, then just run the gui, without trying to install from scratch. It also tries to link .so files again and fails since they already exist. We also get space countdown when there's nothing new to install. On the plus side, `-p` flag works ;)

**Comment by jstayco** [comment]
@Galunid I think this is a good solution. By default we are hiding a lot of output that we previously surfaced.  If you run into problems or need that information you can now run it with -v, -vv, or -vvv for increased verbose modes. We are now also explicitly upgrading pip before Python operations as it can't hurt to update, but it can hurt to be out of date.

There are also now many more guards on the symlink ops in the function create_symlink(). There is also a -s or --skip-space-check option now to explicitly skip that storage check. And since you can skip that manually now, I also added it to the macOS operations.

**Comment by jstayco** [comment]
There are new commands now. I integrated the upgrade.sh stuff into this script because we were really only missing the check for git status in this one. All other functions were already duplicated. So aded that check and the ability to skip git stuff entirely via the -n or --no-git-update switch.

Also consolidated the python stuff since all the pip commands are mostly the same anyway. Should ease maintenance.

**Comment by jstayco** [comment]
@Galunid I think this is as good as it will get and it's ready to merge. If I missed anything, I think it can be handled with individual bug reports after the fact. But I've tested this on a new installation and repairing an existing installation on macOS. Feel like it covers every scenario I could think of.

**Comment by jstayco** [comment]
Minor cleanups done. I'm happy with it. @bmaltais

**Comment by bmaltais** [comment]
@jstayco I tried running the setup script on a M1 Mac and I am getting:

```
bernardmaltais@bernards-mbp kohya_ss % ./setup.sh 
./setup.sh: line 36: conditional binary operator expected
./setup.sh: line 36: syntax error near `"$1"'
./setup.sh: line 36: `  if [[ ! -v "$1" ]] || [[ -z "$1" ]]; then'

```

**Comment by bmaltais** [comment]
I reverted the merge due to MacOS issues running setup. I am trying to see how I can un-close the PR so further work can be done to make sure things work properly for MacOS.

**Comment by bmaltais** [comment]
@jstayco Look like github won't allow to reopen merged PR... lame... sorry about that... can you create a new PR for the MacOS fix? I think it is the only option.
